postgres/attempt.go

```diff
@@ -540,6 +562,59 @@ func (w *worker) requestAttemptsForSpec(req coordinate.AttemptRequest, spec *wor
 	return attempts, err
 }
 
+// chooseMakeFailAttempts finds work units to do for a specific work
+// spec, honoring the work spec's MaxRetries field.  It doesn't make
+// sense to call this with maxRetries as zero; call
+// chooseAndMakeAttempts instead.
+func (w *worker) chooseMakeFailAttempts(
+	tx *sql.Tx,
+	spec *workSpec,
+	numUnits int,
+	maxRetries int,
+	now time.Time,
+	length time.Duration,
+) ([]coordinate.Attempt, error) {
+	var attempts []coordinate.Attempt
+	for len(attempts) < numUnits {
+		moreAttempts, err := w.chooseAndMakeAttempts(
+			tx, spec, numUnits-len(attempts), now, length)
+		if err != nil {
+			return nil, err
+		}
+		if len(moreAttempts) == 0 {
+			return attempts, nil
+		}
+
+		// For each of the (new) attempts, count the number of
+		// existing attempts for the work unit and maybe fail it.
+		// (It might be nice to do this in a batch, maybe moving
+		// this logic inside chooseAndMakeAttempts?)
```

Comment:
Do you have a feel for whether this will have a noticeable impact on performance or not? I don't quite have the cost model loaded into my head, but if a work spec has a non-zero max retry setting, then it looks like getting attempts has gone from "one query to get them all" to "one query to get them all, and then at least one additional extra query for each attempt." But I might be missing other steps that make this change more marginal than I think.